### DISCUSSION TITLE
Add reset function to useFetcher object to support modal usage

### DISCUSSION THREAD

**Discussion by cblauvelt** [remix-run/react-router#10231] [Proposals]
I am using useFetcher to submit form data without navigation within a modal. When a user submits data within the modal and closes the modal I would like clear the fetcher state so when the user re-opens the modal they get the same experience as the first usage.

I can work around this by rendering the modal only when needed which mounts and unmounts the modal component but this seems sub-optimal. Ex:
```jsx
{modalOpen ? (<ModalComponent />) : null }
```

See discussion at remix-run/remix#2749.

**Comment by brophdawg11** [comment]
Yep - I'm on board.  @jacob-ebey and I actually almost [shipped](https://github.com/remix-run/react-router/pull/11985) this as part of v7 so we can dust that off.  At the time we were (probably needlessly) waffling on whether we needed reset or abort or both etc.

Let's just ship `fetcher.reset()` which will abort any in-flight calls.  If a standalone abort mechanism ever becomes super necessary we can revisit `fetcher.abort`.  But in the meantime they can accomplish that with `fetcher.reset` and syncing state.

**Reply by jacob-ebey** [reply]
👍

**Comment by cblauvelt** [comment]
Sounds great. Will this issue show when it's merged into main or a development branch?

**Reply by brophdawg11** [reply]
Not here automatically - you can follow https://github.com/remix-run/react-router/issues/14207.  We usually try to add a comment on the discussion manually but the issue will be the most reliable thing to track.

**Comment by brophdawg11** [comment]
Issue: https://github.com/remix-run/react-router/issues/14207
PR: https://github.com/remix-run/react-router/pull/14206

**Reply by brophdawg11** [reply]
If anyone wants to do some alpha testing, this API is now available in `0.0.0-experimental-7ed11be` 🙌

**Reply by brophdawg11** [reply]
Thanks to the folks below for doing some testing!  I forgot to drop this her, but this API shipped as `unstable_reset` in [7.9.2](https://github.com/remix-run/react-router/blob/main/CHANGELOG.md#fetcher-reset-unstable).

**Comment by bravo-kernel** [comment]
Late to the party but this seems to work (perfect) for my use case.

```ts
useFormFieldEmptied(searchValue, () => fetcher.unstable_reset());
```

Much appreciated folks 💖

**Comment by AndreSilva1993** [comment]
Just tested as well and it works beautifully. Thanks for shipping this!

**Comment by brookslybrand** [comment]
Much love for this feature. Thanks @brophdawg11!

Previously, to accomplish this little UX I'd have to pull in useState and keep another state in sync.



https://github.com/user-attachments/assets/0492f8e2-2115-4dfd-a628-5f214678ca07



Now I can just do this:

```tsx
onChange={() => {
  if (isSuccess) {
    fetcher.unstable_reset();
  }
}}
```

**Comment by brophdawg11** [comment]
Closing - this shipped as unstable in 7.9.2 and stabilized in 7.10.0

**Reply by estenp** [reply]
Does anyone know when and how the API docs get updated? They appear to be a typedoc site - shouldn't it be up to date with the API's in the latest version? Or am I looking in the wrong place?

https://api.reactrouter.com/v7/types/react_router.FetcherWithComponents.html

**Reply by brophdawg11** [reply]
oops - yeah these are unfortunately still a manual process and it looks like we're a bit stale.  I'll get that updated (and hopefully automated 😄 )

Until then, if you want to see the latest, you can clone the `react-router` repo and run `pnpm i; pnpm builld; pnpm run docs;` and the docs will be generated in `public/dev/`

**Reply by brophdawg11** [reply]
this is now updated and automated going forward 👍

**Reply by brophdawg11** [reply]
The updated version of typedoc made a small change to file paths, so the above URL is now https://api.reactrouter.com/v7/types/react-router.FetcherWithComponents.html

**Reply by bravo-kernel** [reply]
#heromaterial 💖
